src/crafting/PotionTypeRecipe.php

```diff
@@ -0,0 +1,58 @@
+<?php
+
+/*
+ *
+ *  ____            _        _   __  __ _                  __  __ ____
+ * |  _ \ ___   ___| | _____| |_|  \/  (_)_ __   ___      |  \/  |  _ \
+ * | |_) / _ \ / __| |/ / _ \ __| |\/| | | '_ \ / _ \_____| |\/| | |_) |
+ * |  __/ (_) | (__|   <  __/ |_| |  | | | | | |  __/_____| |  | |  __/
+ * |_|   \___/ \___|_|\_\___|\__|_|  |_|_|_| |_|\___|     |_|  |_|_|
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * @author PocketMine Team
+ * @link http://www.pocketmine.net/
+ *
+ *
+*/
+
+declare(strict_types=1);
+
+namespace pocketmine\crafting;
+
+use pocketmine\item\Item;
+
+class PotionTypeRecipe implements BrewingRecipe{
+
+	/** @var Item */
+	private $input;
+	/** @var Item */
+	private $ingredient;
+	/** @var Item */
+	private $output;
+
+	public function __construct(Item $input, Item $ingredient, Item $output) {
+		$this->input = clone $input;
+		$this->ingredient = clone $ingredient;
+		$this->output = clone $output;
+	}
```

Comment:
This constructor can use promoted properties.

Reply:
Even if those items are being cloned?

Reply:
Yes. You can keep the assignments in the constructor body for the cloning, and still get to use promoted properties to avoid the property declarations.